### PR TITLE
Turbovec: add vector index C API foundation

### PR DESCRIPTION
**Pull request by dev-nid**
## Summary

Adds the initial standalone `ggml-vector-index` C API foundation.

- Introduces `ggml_vec_index_*` public API surface.
- Adds basic in-memory vector index lifecycle, add, search, remove, contains, write/load coverage from the initial implementation slice.
- Wires the initial vector-index source into the build.
- Adds the first `test-vector-index` coverage for API behavior.

## Test plan

- [ ] `cmake -S . -B build -DGGML_VECTOR_INDEX=ON -DLLAMA_BUILD_TESTS=ON`
- [ ] `cmake --build build --target test-vector-index`
- [ ] `ctest --test-dir build -R '^test-vector-index$' --output-on-failure`

## Related PR

embed: https://github.com/tetherto/qvac/pull/3292

## Original reference
https://github.com/RyanCodrai/turbovec

#### Checks after this PR: 
PR2: https://github.com/dev-nid/qvac-fabric-llm.cpp/pull/1

#### After PR2:
PR3: https://github.com/dev-nid/qvac-fabric-llm.cpp/pull/2

### CODE REVIEW

**Review by gianni-cor — CHANGES_REQUESTED** [review]
I reviewed this foundation PR in isolation (including the public C API, add/remove/search behavior, the v1 loader/writer, CMake wiring, and the smoke test). The normal smoke test passes, but there are several edge cases and API/integration concerns that should be addressed before merging. Could you please take a look at the following?

## Correctness and robustness

### 1. Validate the snapshot size before allocating from its header

`ggml/src/ggml-vector-index.cpp:494-516`

The loader trusts `dim` and `n_vectors` from the 16-byte header and immediately resizes the vector, ID array, and hash map. It only discovers a truncated payload later, while reading the data. A tiny malformed file can therefore request several gigabytes of memory and may cause severe memory pressure or terminate the process instead of being rejected cheaply.

Suggested fix: determine the file size first, calculate the exact expected size with checked arithmetic (`16 + N*D*sizeof(float) + N*sizeof(uint64_t)`), and reject mismatches or counts outside supported API limits before constructing/resizing the index. Please add truncated-header and oversized-count tests.

### 2. Reserve `UINT64_MAX` instead of accepting it as a user ID

`ggml/src/ggml-vector-index.cpp:198-209`

Search uses `UINT64_MAX` to mark padded result slots, but `ggml_vec_index_add` currently accepts the same value as a real external ID. For example, inserting one vector with ID `UINT64_MAX` and searching with `k=2` returns `UINT64_MAX` for both the real hit and the padded slot. A caller using the documented sentinel cannot distinguish them.

Suggested fix: reject `UINT64_MAX` in both `add` and `load`, document that it is reserved, and add a regression test covering both paths.

### 3. Keep search scores finite and preserve descending order

`ggml/src/ggml-vector-index.cpp:308-314, 351-358`

The dot product accumulates in `float` without input or result validation. Even finite inputs can overflow: storing `-FLT_MAX`, querying with `FLT_MAX`, and requesting `k=2` produces `[-INFINITY, -FLT_MAX]`. The padded score is then greater than the real score, contradicting the API guarantee that each row is sorted descending. NaN values are more serious because they also invalidate the heap comparator's strict ordering assumptions.

Suggested fix: reject non-finite vector/query components, accumulate in `double`, and either clamp the final score to `[-FLT_MAX, FLT_MAX]` or return an error if it is not representable. Add tests for finite overflow and explicit NaN/Inf inputs.

### 4. Prevent the public length from overflowing `int`

`ggml/src/ggml-vector-index.cpp:552-554`

The on-disk count is `uint32_t` and the in-memory containers use `size_t`, but `ggml_vec_index_len` casts the size directly to `int`. An index above `INT_MAX` therefore returns a negative or truncated length even though the loader/add paths currently permit that state.

Suggested fix: because the public API returns `int`, reject additions and snapshots that would exceed `INT_MAX` (preferable), or change the public API to a wider count type before declaring it stable.

### 5. Use an isolated temporary file in the smoke test

`tests/test-vector-index.cpp:141-171`

Every test process uses `temp_directory_path()/ggml-vector-index-test.tvim`. Concurrent runs can truncate, load, or remove each other's snapshot, making the test flaky on shared runners or when multiple worktrees are tested at once.

Suggested fix: generate a unique filename using the PID plus a random/counter suffix and clean it up with an RAII guard.

## API and integration concerns

### 6. The “final API” statement does not match the stacked series

`ggml/include/ggml-vector-index.h:5-13, 33-51`

This header says downstream bindings can depend on the API as final, but the next stacked PR changes public behavior and types: this PR accepts every bit width in `[1, 32]`, while the later implementation supports a narrower set, and the public error enum also changes. Publishing the current contract as final creates avoidable source/behavior compatibility risk between consecutive PRs.

Suggested fix: either introduce the final stacked public header and semantics in this foundation PR, or remove the final/backward-compatibility claim and explicitly mark the API experimental until the stack lands.

### 7. The implementation is not actually standalone

`ggml/src/CMakeLists.txt:188-210`, `tests/CMakeLists.txt:147-154`

The implementation is compiled unconditionally into `ggml-base`, the header is installed as a core ggml header, and the test links `ggml`; there is no opt-in option or separate vector-index target. This does not match the PR description's “standalone” framing and temporarily expands every ggml build/ABI even for consumers that do not use the feature.

Suggested fix: bring the opt-in `GGML_VECTOR_INDEX` option and separate `ggml-vector-index` target into this foundation PR, or reword the scope clearly if unconditional core integration is intentional.

Please take a look at these points. The first four affect accepted inputs or untrusted snapshot handling and should be resolved rather than deferred to later optimization work.

**Comment by dev-nid** [comment]
> I reviewed this foundation PR in isolation (including the public C API, add/remove/search behavior, the v1 loader/writer, CMake wiring, and the smoke test). The normal smoke test passes, but there are several edge cases and API/integration concerns that should be addressed before merging. Could you please take a look at the following?
> 
> ## Correctness and robustness
> ### 1. Validate the snapshot size before allocating from its header
> `ggml/src/ggml-vector-index.cpp:494-516`
> 
> The loader trusts `dim` and `n_vectors` from the 16-byte header and immediately resizes the vector, ID array, and hash map. It only discovers a truncated payload later, while reading the data. A tiny malformed file can therefore request several gigabytes of memory and may cause severe memory pressure or terminate the process instead of being rejected cheaply.
> 
> Suggested fix: determine the file size first, calculate the exact expected size with checked arithmetic (`16 + N*D*sizeof(float) + N*sizeof(uint64_t)`), and reject mismatches or counts outside supported API limits before constructing/resizing the index. Please add truncated-header and oversized-count tests.
> 
> ### 2. Reserve `UINT64_MAX` instead of accepting it as a user ID
> `ggml/src/ggml-vector-index.cpp:198-209`
> 
> Search uses `UINT64_MAX` to mark padded result slots, but `ggml_vec_index_add` currently accepts the same value as a real external ID. For example, inserting one vector with ID `UINT64_MAX` and searching with `k=2` returns `UINT64_MAX` for both the real hit and the padded slot. A caller using the documented sentinel cannot distinguish them.
> 
> Suggested fix: reject `UINT64_MAX` in both `add` and `load`, document that it is reserved, and add a regression test covering both paths.
> 
> ### 3. Keep search scores finite and preserve descending order
> `ggml/src/ggml-vector-index.cpp:308-314, 351-358`
> 
> The dot product accumulates in `float` without input or result validation. Even finite inputs can overflow: storing `-FLT_MAX`, querying with `FLT_MAX`, and requesting `k=2` produces `[-INFINITY, -FLT_MAX]`. The padded score is then greater than the real score, contradicting the API guarantee that each row is sorted descending. NaN values are more serious because they also invalidate the heap comparator's strict ordering assumptions.
> 
> Suggested fix: reject non-finite vector/query components, accumulate in `double`, and either clamp the final score to `[-FLT_MAX, FLT_MAX]` or return an error if it is not representable. Add tests for finite overflow and explicit NaN/Inf inputs.
> 
> ### 4. Prevent the public length from overflowing `int`
> `ggml/src/ggml-vector-index.cpp:552-554`
> 
> The on-disk count is `uint32_t` and the in-memory containers use `size_t`, but `ggml_vec_index_len` casts the size directly to `int`. An index above `INT_MAX` therefore returns a negative or truncated length even though the loader/add paths currently permit that state.
> 
> Suggested fix: because the public API returns `int`, reject additions and snapshots that would exceed `INT_MAX` (preferable), or change the public API to a wider count type before declaring it stable.
> 
> ### 5. Use an isolated temporary file in the smoke test
> `tests/test-vector-index.cpp:141-171`
> 
> Every test process uses `temp_directory_path()/ggml-vector-index-test.tvim`. Concurrent runs can truncate, load, or remove each other's snapshot, making the test flaky on shared runners or when multiple worktrees are tested at once.
> 
> Suggested fix: generate a unique filename using the PID plus a random/counter suffix and clean it up with an RAII guard.
> 
> ## API and integration concerns
> ### 6. The “final API” statement does not match the stacked series
> `ggml/include/ggml-vector-index.h:5-13, 33-51`
> 
> This header says downstream bindings can depend on the API as final, but the next stacked PR changes public behavior and types: this PR accepts every bit width in `[1, 32]`, while the later implementation supports a narrower set, and the public error enum also changes. Publishing the current contract as final creates avoidable source/behavior compatibility risk between consecutive PRs.
> 
> Suggested fix: either introduce the final stacked public header and semantics in this foundation PR, or remove the final/backward-compatibility claim and explicitly mark the API experimental until the stack lands.
> 
> ### 7. The implementation is not actually standalone
> `ggml/src/CMakeLists.txt:188-210`, `tests/CMakeLists.txt:147-154`
> 
> The implementation is compiled unconditionally into `ggml-base`, the header is installed as a core ggml header, and the test links `ggml`; there is no opt-in option or separate vector-index target. This does not match the PR description's “standalone” framing and temporarily expands every ggml build/ABI even for consumers that do not use the feature.
> 
> Suggested fix: bring the opt-in `GGML_VECTOR_INDEX` option and separate `ggml-vector-index` target into this foundation PR, or reword the scope clearly if unconditional core integration is intentional.
> 
> Please take a look at these points. The first four affect accepted inputs or untrusted snapshot handling and should be resolved rather than deferred to later optimization work.

Addressed

File: ggml/src/CMakeLists.txt [hunk 1]
```diff
@@ -514,6 +514,36 @@ foreach (target ggml-base ggml)
...
+
+    if (DEFINED MATH_LIBRARY)
+        target_link_libraries(ggml-vector-index PRIVATE ${MATH_LIBRARY})
+    elseif (NOT WIN32 AND NOT DEFINED ENV{ONEAPI_ROOT})
+        target_link_libraries(ggml-vector-index PRIVATE m)
+    endif()
+
+    if (BUILD_SHARED_LIBS OR GGML_BACKEND_DL)
+        set_target_properties(ggml-vector-index PROPERTIES POSITION_INDEPENDENT_CODE ON)
+        target_compile_definitions(ggml-vector-index PRIVATE GGML_BUILD)
```
Comment (gianni-cor) [inline ggml/src/CMakeLists.txt]: Windows shared builds still need `GGML_SHARED` here. `GGML_API` only expands to `__declspec(dllexport)` when both `GGML_SHARED` and `GGML_BUILD` are defined, while this target currently defines only the latter. Please propagate `GGML_SHARED` publicly for shared builds, matching the existing `ggml`/`ggml-base` handling, and verify it with a Windows shared build of `test-vector-index`.

File: ggml/CMakeLists.txt [hunk 2]
```diff
@@ -123,6 +123,7 @@ option(GGML_STATIC "ggml: static link libraries"                     OFF)
 option(GGML_NATIVE "ggml: optimize the build for the current system" ${GGML_NATIVE_DEFAULT})
 option(GGML_LTO    "ggml: enable link time optimization"             OFF)
 option(GGML_CCACHE "ggml: use ccache if available"                   ON)
+option(GGML_VECTOR_INDEX "ggml: build standalone vector index library" OFF)
```
Comment (gianni-cor) [inline ggml/CMakeLists.txt:126]: Because this option defaults to `OFF` and the existing workflows do not pass `-DGGML_VECTOR_INDEX=ON`, the current green CI checks skip both the new library and `test-vector-index`. Please enable it in at least one Linux job and one Windows shared job (or add a focused matrix) so this code and the DLL exports are exercised.

**Review by gianni-cor — COMMENTED** [review]
Two remaining integration issues are noted inline.

File: ggml/src/CMakeLists.txt (hunk 1 above)
Comment (dev-nid) [reply ggml/src/CMakeLists.txt]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/fcb520f50e7afa8f1d6d41474065c2496881bbf4

File: ggml/CMakeLists.txt (hunk 2 above)
Comment (dev-nid) [reply ggml/CMakeLists.txt:126]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/fcb520f50e7afa8f1d6d41474065c2496881bbf4

File: ggml/include/ggml-vector-index.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+typedef struct ggml_vec_index ggml_vec_index_t;
+
+// Error codes returned from int-valued APIs. 0 = OK. Negative = failure.
+// `_remove` is the exception: it returns 1 on removal and 0 on miss.
+enum ggml_vec_index_error {
+    GGML_VEC_INDEX_OK            = 0,
+    GGML_VEC_INDEX_E_INVALID_ARG = -2,
+    GGML_VEC_INDEX_E_DUPLICATE   = -3,
+    GGML_VEC_INDEX_E_IO          = -4,
+    GGML_VEC_INDEX_E_OOM         = -7,
```
Comment (aegioscy) [inline ggml/include/ggml-vector-index.h]: why are these numbers so weird? are these GGML numbers -4, -7? super strange for enums

**Review by aegioscy — COMMENTED** [review]
just address the comment

File: ggml/include/ggml-vector-index.h (hunk 3 above)
Comment (dev-nid) [reply ggml/include/ggml-vector-index.h]: They are not GGML-wide numbers. The gaps came from the stacked follow-up branches, where -5, -6, and -8 are used for persistence/compaction errors. Agreed that this looks odd in this PR subset, so I’ll renumber the visible errors sequentially and keep the later additions sequential in the stack.

Comment (dev-nid) [reply ggml/include/ggml-vector-index.h]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/7744b4c4dd6e296611873cdd1a0810adf4cb69d3

**Comment by dev-nid** [comment]
> just address the comment

Addressed

**Review by gianni-cor — COMMENTED** [review]
Inline findings from the TurboVec series review.

File: ggml/src/ggml-vector-index.cpp [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+            return GGML_VEC_INDEX_E_INTERNAL;
+        }
+        if (idx->data.size() != n * dim_sz) {
+            return GGML_VEC_INDEX_E_INTERNAL;
+        }
+
+        const std::filesystem::path dst_path(path);
+        const std::filesystem::path tmp_path = make_tmp_path(dst_path);
+
+        std::unique_ptr<std::FILE, decltype(&std::fclose)> f(open_exclusive(tmp_path), &std::fclose);
```
Comment (gianni-cor) [inline ggml/src/ggml-vector-index.cpp]: This deleter type triggers GCC's `-Wignored-attributes` warning because attributes attached to `fclose` are discarded when the function type is used as a template argument. The Linux CPU workflow enables fatal warnings, so enabling `GGML_VECTOR_INDEX` makes that build fail. A small stateless deleter struct that calls `std::fclose` avoids using the attributed function type as a template parameter. Could you take a look and switch this to an attribute-safe deleter?

File: ggml/ggml.pc.in [hunk 5]
```diff
@@ -0,0 +1,10 @@
+prefix=@GGML_PC_PREFIX@
+exec_prefix=${prefix}
+libdir=@GGML_PC_LIBDIR@
+includedir=@GGML_PC_INCLUDEDIR@
+
+Name: ggml
+Description: Tensor library for machine learning
+Version: @GGML_VERSION@
+Libs: -L${libdir} -lggml -lggml-base
```
Comment (gianni-cor) [inline ggml/ggml.pc.in:9]: The pkg-config metadata exposes only `-lggml -lggml-base`. Static consumers also need the C++ runtime and the private libraries pulled in by the vector-index implementation; otherwise a C consumer using `pkg-config --static ggml` gets unresolved C++ and backend symbols. Please consider generating `Libs.private` from the target's private link dependencies (or publishing the vector-index component separately) and add a static pkg-config consumer test. Could you take a look?

File: ggml/src/ggml-vector-index.cpp [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+        CloseHandle(handle);
+        return nullptr;
+    }
+    std::FILE * file = _fdopen(fd, "wb");
+    if (file == nullptr) {
+        _close(fd);
+    }
+    return file;
+#else
+    const int fd = open(path.c_str(), O_WRONLY | O_CREAT | O_EXCL, 0666);
```
Comment (gianni-cor) [inline ggml/src/ggml-vector-index.cpp]: This creates the replacement snapshot with mode `0666`, even when the existing destination is more restrictive. Until the rename, a permissive umask can therefore expose vector data more broadly than the file being replaced. Please create the temporary file restrictively and `fchmod` it to the destination's mode before publishing it; for a brand-new destination, apply the intended `0666 & ~umask` policy explicitly. Could you take a look?

Comment (dev-nid) [reply ggml/src/ggml-vector-index.cpp]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/4d1289d97b24b3c5c980675083fcf42480edd12f

File: ggml/ggml.pc.in (hunk 5 above)
Comment (dev-nid) [reply ggml/ggml.pc.in:9]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/985f342955f7175b9b6a6ab2a12242a42cd6c6c6

File: ggml/src/ggml-vector-index.cpp (hunk 4 above)
Comment (dev-nid) [reply ggml/src/ggml-vector-index.cpp]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/4d1289d97b24b3c5c980675083fcf42480edd12f

File: ggml/src/ggml-vector-index.cpp [hunk 7]
```diff
@@ -0,0 +1,930 @@
...
+        base_slot = idx->slot_to_id.size();
+        if (n_sz > std::numeric_limits<size_t>::max() - base_slot) {
+            return GGML_VEC_INDEX_E_INVALID_ARG;
+        }
+        const size_t new_slots = base_slot + n_sz;
+        if (dim_sz != 0 && new_slots > std::numeric_limits<size_t>::max() / dim_sz) {
+            return GGML_VEC_INDEX_E_INVALID_ARG;
+        }
+
+        idx->data.reserve(new_slots * dim_sz);
```
Comment (jesusmb1995) [inline ggml/src/ggml-vector-index.cpp]: This can rehash on every call, is there a better solution? (e.g. can we pre-allocate once at the caller?)

Comment (dev-nid) [reply ggml/src/ggml-vector-index.cpp]: Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/3efc329564663e9add7448e0e4cf2eaa2b4e75d5

**Comment by gianni-cor** [comment]
permission change is not crash-durable

The snapshot is written and synced, then fchmod() changes its permissions before rename. If the system crashes after fchmod() but before another file sync, the file contents may be durable while its mode change is not. That can leave a replacement snapshot with unexpected permissions after recovery.

The fix is to sync the temporary file again after fchmod() succeeds, then perform the atomic rename and parent-directory sync.

**Comment by dev-nid** [comment]
> permission change is not crash-durable
> 
> The snapshot is written and synced, then fchmod() changes its permissions before rename. If the system crashes after fchmod() but before another file sync, the file contents may be durable while its mode change is not. That can leave a replacement snapshot with unexpected permissions after recovery.
> 
> The fix is to sync the temporary file again after fchmod() succeeds, then perform the atomic rename and parent-directory sync.

Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/d17295d1b6954b28e4ed399683e21a52b04141f3

**Comment by gianni-cor** [comment]
A successful snapshot write can currently create a file that the same library will refuse to load.

`ggml_vec_index_load()` applies the checked `kMaxSnapshotBytes` limit (4 GiB) before allocating, but `ggml_vec_index_write()` validates dimensions/counts without applying that same serialized-size limit. The public in-memory API can therefore accept an index whose v1 snapshot is larger than 4 GiB; `write()` succeeds, while a later `load()` rejects the otherwise-valid snapshot solely due to the loader cap.

This makes persistence non-round-trippable for a supported in-memory state. For example, a valid `dim=4096` index with 262,144 rows exceeds the cap once vectors and IDs are serialized.

Suggested fix: calculate the checked serialized size during `write()` and reject sizes above the same limit before creating the temporary file; document the persistence size limit and add a regression that verifies a too-large in-memory index fails to write rather than producing an unloadable file. Alternatively, raise/remove the loader limit while keeping equivalent bounded-allocation protections.

Could you please take a look and ensure write/load have a consistent supported-size contract?

**Comment by dev-nid** [comment]
> A successful snapshot write can currently create a file that the same library will refuse to load.
> 
> `ggml_vec_index_load()` applies the checked `kMaxSnapshotBytes` limit (4 GiB) before allocating, but `ggml_vec_index_write()` validates dimensions/counts without applying that same serialized-size limit. The public in-memory API can therefore accept an index whose v1 snapshot is larger than 4 GiB; `write()` succeeds, while a later `load()` rejects the otherwise-valid snapshot solely due to the loader cap.
> 
> This makes persistence non-round-trippable for a supported in-memory state. For example, a valid `dim=4096` index with 262,144 rows exceeds the cap once vectors and IDs are serialized.
> 
> Suggested fix: calculate the checked serialized size during `write()` and reject sizes above the same limit before creating the temporary file; document the persistence size limit and add a regression that verifies a too-large in-memory index fails to write rather than producing an unloadable file. Alternatively, raise/remove the loader limit while keeping equivalent bounded-allocation protections.
> 
> Could you please take a look and ensure write/load have a consistent supported-size contract?

Addressed in https://github.com/tetherto/qvac-fabric-llm.cpp/pull/192/commits/28153a07058b1230fac5bd4b4252acae5434e9f8
write() now uses the same checked v1 snapshot-size preflight as load(), rejects snapshots over the 4 GiB persistence limit before creating a temp file, and the limit is documented with regression coverage for the oversized case.
